GoogleForCreators/web-stories-wp
# Page Layouts: Keyboard Navigation

## Summary

Refactors the page layout panel a little bit to allow our virtualized list to be less nested and use the useRovingIndex util to handle arrow navigation.

**Below is revised as of Jan 28, 2021**
## Relevant Technical Choices

- Implement a secondary virtualizedGrid for columns (like this: https://codesandbox.io/s/github/tannerlinsley/react-virtual/tree/master/examples/dynamic) to remove some of the panel's nesting and make dom easier to traverse. 
- Refactors grid and places the interactive `button` type div to be the wrapper of each layout.  This replaces the `listitem` element type, to me button makes more sense since it tells us that you can click it? 
- I had to move all of the logic for focus of a page layout to the parent and pass it in, making `PageLayout` purely presentational _because_ the grid is virtualized which means whenever it is triggered to go get another row the DOM is updating in the background. It looks the same, but it is not. This is why focus was disappearing before! 
- I also moved the dialog to confirm above the container that holds the virtualized layouts, this is important placement because it triggers the `useFocusOut` hook that allows us to refocus when the dialog is closed. 
- Also because of this refocus, when you come back to a page layout after dealing with the dialog you have to keyDown an arrow twice to actually move because the first is how the virtualized list is grappling with its current position. 


## To-do

- Karma tests once Zach's initial structure for testing is in place
- Extract the code here that I've added that controls the keyboard nav with virtualized grid to a hook - since this same bug exists in the text sets I figure I'll do this there.

## User-facing changes

- Users should now be able to navigate the page layouts with arrow keys

## Testing Instructions

- Should be able to navigate page layouts with arrow keys 
- Tabbing from a page layout will send you to the canvas 
- 'Enter' when hit from a page layout should either swap the page layout if the canvas is empty or open a dialog. If a dialog enters you should be able to tab to the buttons on it. When you close the dialog the focus should return to the page layout you were on when you entered the dialog. 
- Confirm the same holds true for RTL
- Confirm screen reader still gives accurate information about page layouts
- Confirm scroll is still good to go

---



Addresses #5876
